### DISCUSSION TITLE
Should the plugin author use `VOID_TRIGGER` explicitly or not?

### DISCUSSION THREAD

**Discussion by davy-tw** [ChildishGhost/einstein#1] [API Design]
Originally from comment: https://github.com/ChildishGhost/einstein/commit/4b18edd5814efafda4f257ce8f48a08905183e6d#r47785876
> @xatier: thought: should the plugin author use VOID_TRIGGER explicitly or not? Let's make this consistent.

**Comment by davy-tw** [comment]
IMO, `VOID_TRIGGER` should represent "no triggers needed" which has special meaning.
Since it is given by empty string `''` now, I think we can change the value to unique `Symbol` and skip empty-string triggers in system.

That is, developers need to use `VOID_TRIGGER` explicitly and not to use empty-string as alternative.

**Reply by xatier** [reply]
sounds good to me.

**Comment by xatier** [comment]
Sync'd offline.

1. We should provide both explicit and implicit usage of `VOID_TRIGGER`.
2. The implicit behavior will be documented in the parameter type polymorphism.
3. `VOID_TRIGGER` will become a `Symbol`, so plugin authors can explicitly specify both plugin's trigger _and_ `VOID_TRIGGER`. This can be helpful when the plugin author would like to register some help menu which is intended to be displayed in all cases.
4. When the explicit triggers are specified, we shall limit the search results within the engines that provides those triggers.
